stbt/android.py

```diff
@@ -0,0 +1,784 @@
+        screenshot (scale and/or rotate it) to match the screenshots from your
+        main video-capture method as closely as possible, as specified by the
+        `CoordinateSystem`.
+
+        :returns: A `stbt.Frame`, that is, an image in OpenCV format. Note that
+            the ``time`` attribute won't be very accurate (probably to <0.5s or
+            so).
+        """
+
+        for attempt in range(1, 4):
+            timestamp = time.time()
+            data = (self.adb(["shell", "screencap", "-p"],
+                             timeout_secs=60, capture_output=True)
+                    .replace("\r\n", "\n"))
+            img = cv2.imdecode(
+                numpy.asarray(bytearray(data), dtype=numpy.uint8),
+                cv2.CV_LOAD_IMAGE_COLOR)
+            if img is None:
+                logging.warning(
+                    "AdbDevice.get_frame: Failed to get screenshot "
+                    "via ADB (attempt %d/10)\n"
+                    "Length of data: %d", attempt, len(data))
+            else:
+                break
+        else:
+            raise RuntimeError(
+                "Failed to capture screenshot from android device")
+
+        w, h = img.shape[1], img.shape[0]
+
+        if self.coordinate_system == CoordinateSystem.ADB_NATIVE:
+            pass
+        elif self.coordinate_system == CoordinateSystem.ADB_720P:
+            # Resize to 720p preserving orientation
+            if w > h:
+                img = cv2.resize(img, (1280, 720))
+            else:
+                img = cv2.resize(img, (720, 1280))
+        elif self.coordinate_system == CoordinateSystem.CAMERA_720P:
+            # Resize to 720p landscape for compatibility with screenshots from
+            # `stbt.get_frame` with the Stb-tester CAMERA.
+            if w > h:
+                img = cv2.resize(img, (1280, 720))
+            else:
+                img = numpy.rot90(cv2.resize(img, (720, 1280)))
+        else:
+            raise NotImplementedError(
+                "AdbDevice.get_frame not implemented for %s. "
+                "Use a separate AdbDevice instance with "
+                "coordinate_system=CoordinateSystem.ADB_NATIVE"
+                % self.coordinate_system)
+
+        return stbt.Frame(img, time=timestamp)
+
+    def press(self, key):
+        """Send a keypress.
+
+        :param str key: An Android keycode as listed in
+            https://developer.android.com/reference/android/view/KeyEvent.html
+            (you can specify it with or without the "KEYCODE_" prefix).
+            Particularly useful key codes are "HOME" and "BACK", which are
+            physical buttons on some phones so you can't hit them with
+            `AdbDevice.tap`.
+        """
+        # "adb shell input keyevent xxx" always returns success, so we need to
+        # validate key names.
+        if key not in _KEYCODES and "KEYCODE_" + key not in _KEYCODES:
+            raise ValueError("Unknown key code %r" % (key,))
```

Comment:
`press("KEY_OK")` will fail. It'll look up `KEYCODE_KEY_OK`.

I would say require the `KEYCODE_` prefix for now.  That way we can sort out a mapping for our standard keycodes in the future without clashes or backwards compatibility concerns.

Is there a reason you need to check the list before forwarding it on to ADB?  Presumably `adb` will fail if the key is missing?